src/main/java/com/aws/greengrass/certificatemanager/CertificateManager.java

```diff
@@ -114,30 +119,55 @@ public String getCaPassPhrase() {
      * @param cb  Certificate consumer
      * @throws KeyStoreException if unable to access KeyStore
      * @throws CsrProcessingException if unable to process CSR
+     * @throws RuntimeException for any unknown runtime error
      */
-    @SuppressWarnings("PMD.AvoidCatchingGenericException")
+    @SuppressWarnings({"PMD.AvoidCatchingGenericException", "PMD.ExceptionAsFlowControl"})
     public void subscribeToServerCertificateUpdates(@NonNull String csr, @NonNull Consumer<X509Certificate> cb)
             throws KeyStoreException, CsrProcessingException {
-        // BouncyCastle can throw RuntimeExceptions, and unfortunately it is not easy to detect
-        // bad input beforehand. For now, just catch and re-throw a CsrProcessingException
+        // deduplicate subscriptions with the same callback
         try {
-            PKCS10CertificationRequest pkcs10CertificationRequest =
-                    CertificateHelper.getPKCS10CertificationRequestFromPem(csr);
-            JcaPKCS10CertificationRequest jcaRequest = new JcaPKCS10CertificationRequest(pkcs10CertificationRequest);
-            CertificateGenerator certificateGenerator = new ServerCertificateGenerator(
-                    jcaRequest.getSubject(), jcaRequest.getPublicKey(), cb, certificateStore);
-
-            // Add certificate generator to monitors first in order to avoid missing events
-            // that happen while the initial certificate is being generated.
-            certExpiryMonitor.addToMonitor(certificateGenerator);
-            cisShadowMonitor.addToMonitor(certificateGenerator);
-
-            certificateGenerator.generateCertificate(connectivityInfoProvider::getCachedHostAddresses);
-        } catch (KeyStoreException e) {
-            logger.atError().setCause(e).log("unable to subscribe to certificate update");
+            serverCertSubscriptions.compute(cb, (k, v) -> {
+                // A subscription already exists, we will replace it so that a new certificate is generated immediately
+                if (v != null) {
+                    unsubscribeFromServerCertificateUpdates(cb);
```

Comment:
Lets not modify the map while in the compute function. 

`Attempts to compute a mapping for the specified key and its current mapped value (or null if there is no current mapping). The entire method invocation is performed atomically. Some attempted update operations on this map by other threads may be blocked while computation is in progress, so the computation should be short and simple, and must not attempt to update any other mappings of this Map.`

Reply:
Done.